File: backend/app/gazetteer.py
"""Bundled US terrain gazetteer (USGS GNIS, public domain).

This is the default search backend. It ships with the app so search works
with no third-party service, no API key, and no request budget - see
search.py for why that matters.

The index is a gzipped TSV built by scripts/build_gnis_gazetteer.py, loaded
lazily on the first query and kept in memory afterwards (~190k features,
tens of MB resident). Lookup is a linear scan with a small scoring function;
at this size that lands well under a typing-latency budget, and it avoids
shipping a search-engine dependency for one text box.
"""

import csv
import gzip
import sys
import threading
from pathlib import Path

DATA = Path(__file__).resolve().parent / "data" / "gnis_terrain.tsv.gz"

# Search-result priority when scores tie. Mirrors the build script's class
# ordering; a query that hits a summit and a valley should show the summit.
CLASS_RANK = {
    "Summit": 0, "Range": 1, "Gap": 2, "Ridge": 3, "Pillar": 4,
    "Glacier": 5, "Crater": 6, "Arch": 7, "Cliff": 8, "Basin": 9,
    "Valley": 10, "Falls": 11, "Slope": 12, "Bench": 13, "Flat": 14,
    "Lava": 15, "Isthmus": 16,
}

# (name, lower_name, feature_class, state, county, lat, lon)
_rows: list[tuple[str, str, str, str, str, float, float]] | None = None
_lock = threading.Lock()


def available() -> bool:
    return DATA.exists()


def _load() -> list[tuple[str, str, str, str, str, float, float]]:
    global _rows
    if _rows is not None:
        return _rows
    with _lock:
        if _rows is not None:  # another thread won the race
            return _rows
        rows = []
        with gzip.open(DATA, "rt", encoding="utf-8", newline="") as fh:
            for rec in csv.reader(fh, delimiter="\t"):
                if len(rec) != 6:
                    continue
                name, cls, state, county, lat, lon = rec
                try:
                    lat_f, lon_f = float(lat), float(lon)
                except ValueError:
                    continue
                # class/state/county repeat across ~190k rows from a few
                # thousand distinct values; interning collapses the copies.
                rows.append(
                    (
                        name,
                        name.lower(),
                        sys.intern(cls),
                        sys.intern(state),
                        sys.intern(county),
                        lat_f,
                        lon_f,
                    )
                )
        _rows = rows
    return _rows


def _score(needle: str, hay: str) -> int:
    """Lower is better; None-equivalent is -1 meaning no match.

    Ranked so that "capitol" puts Capitol Peak above Capitol Reef Valley, and
    an exact "Little Bear Peak" beats "Little Bear Peak Southwest Ridge".
    """
    if hay == needle:
        return 0
    if hay.startswith(needle):
        return 1
    # word-boundary hit: "bear peak" should match "Little Bear Peak"
    if f" {needle}" in hay:
        return 2
    if needle in hay:
        return 3
    return -1


# Names GNIS files under a different official name than people search for.
# GNIS carries only the current official name per feature, so a search for a
# widely-used alternative finds nothing (or, worse, finds an unrelated minor
# feature that happens to share the word). Kept deliberately tiny - this is
# for names where the mismatch is nationally known, not a synonym dictionary.
ALIASES = {
    "denali": "mount mckinley",
}


def search(q: str, limit: int = 8) -> list[dict]:
    needle = " ".join(q.lower().split())
    if not needle:
        return []
    needle = ALIASES.get(needle, needle)
    rows = _load()

    # Group by name so the result list can show variety. GNIS has 200-plus
    # Bald Mountains and six Capitol Peaks; ranking purely by score buries
    # everything else under near-identical rows.
    groups: dict[str, list[tuple]] = {}
    for name, lower, cls, state, county, lat, lon in rows:
        s = _score(needle, lower)
        if s < 0:
            continue
        # Prefer the shorter name among equal-quality matches: "Bear Peak"
        # over "Bear Peak Trailhead Overlook".
        groups.setdefault(lower, []).append(
            (s, CLASS_RANK.get(cls, 99), len(name), name, cls, state, county, lat, lon)
        )

    for g in groups.values():
        g.sort(key=lambda h: h[:3])
    ordered = sorted(groups.values(), key=lambda g: g[0][:3])

    # Round-robin: every distinct name contributes its best hit before any
    # name contributes a second. Quality ordering still governs within a round.
    out: list[dict] = []
    seen: set[tuple] = set()
    for depth in range(max((len(g) for g in ordered), default=0)):
        for g in ordered:
            if depth >= len(g):
                continue
            _, _, _, name, cls, state, county, lat, lon = g[depth]
            # Same feature mapped twice (GNIS carries variant records)
            # collapses; genuinely distinct peaks sharing a name do not,
            # which is why county rides along in the label.
            key = (name.lower(), round(lat, 2), round(lon, 2))
            if key in seen:
                continue
            seen.add(key)
            region = ", ".join(p for p in (county, state) if p)
            out.append(
                {
                    "name": f"{name}, {region}" if region else name,
                    "lat": lat,
                    "lon": lon,
                    "type": cls.lower(),
                }
            )
            if len(out) >= limit:
                return out
    return out

File: frontend/src/direct/gazetteer.ts
/**
 * The GNIS terrain gazetteer, searched in the browser.
 *
 * A port of backend/app/gazetteer.py. The index is the same gzipped TSV the
 * backend reads - it is a static asset, so nothing about searching it needs a
 * server. Moving it here removes the last routine reason for the app to call
 * a backend at all, and has the side benefit that search keeps working
 * offline once the file is cached.
 *
 * It is fetched lazily on the first query (about 2.9 MB gzipped, ~190k
 * features) and decompressed with DecompressionStream, which is native in
 * every browser this app targets.
 */

export interface GazetteerHit {
  name: string
  lat: number
  lon: number
  type: string
}

const CLASS_RANK: Record<string, number> = {
  Summit: 0, Range: 1, Gap: 2, Ridge: 3, Pillar: 4,
  Glacier: 5, Crater: 6, Arch: 7, Cliff: 8, Basin: 9,
  Valley: 10, Falls: 11, Slope: 12, Bench: 13, Flat: 14,
  Lava: 15, Isthmus: 16,
}

// Names GNIS files differently from what people search for. Kept tiny on
// purpose - this is not a synonym dictionary.
const ALIASES: Record<string, string> = {
  denali: 'mount mckinley',
}

interface Row {
  name: string
  lower: string
  cls: string
  state: string
  county: string
  lat: number
  lon: number
}

let rows: Row[] | null = null
let loading: Promise<Row[]> | null = null

export function isLoaded(): boolean {
  return rows !== null
}

async function load(url: string): Promise<Row[]> {
  if (rows) return rows
  if (loading) return loading
  loading = (async () => {
    const res = await fetch(url)
    if (!res.ok) throw new Error(`Gazetteer fetch failed (${res.status})`)

    // Hosts disagree about a .gz asset. Some (Vite's dev server, and most
    // static hosts) send Content-Encoding: gzip, so the browser has already
    // decompressed it by the time we see the body; others serve the bytes
    // verbatim. Sniffing the gzip magic number handles both without
    // depending on how a particular deploy is configured.
    const buf = await res.arrayBuffer()
    const head = new Uint8Array(buf.slice(0, 2))
    const stillGzipped = head[0] === 0x1f && head[1] === 0x8b
    const text = stillGzipped
      ? await new Response(
          new Blob([buf]).stream().pipeThrough(new DecompressionStream('gzip')),
        ).text()
      : new TextDecoder().decode(buf)

    const out: Row[] = []
    let start = 0
    while (start < text.length) {
      let end = text.indexOf('\n', start)
      if (end === -1) end = text.length
      const line = text.slice(start, end)
      start = end + 1
      if (!line) continue
      const p = line.split('\t')
      if (p.length !== 6) continue
      const lat = +p[4]
      const lon = +p[5]
      if (!isFinite(lat) || !isFinite(lon)) continue
      out.push({ name: p[0], lower: p[0].toLowerCase(), cls: p[1], state: p[2], county: p[3], lat, lon })
    }
    rows = out
    return out
  })()
  return loading
}

/** Lower is better; -1 means no match. Mirrors the Python scorer. */
function score(needle: string, hay: string): number {
  if (hay === needle) return 0
  if (hay.startsWith(needle)) return 1
  if (hay.includes(' ' + needle)) return 2
  if (hay.includes(needle)) return 3
  return -1
}

export async function search(q: string, url: string, limit = 8): Promise<GazetteerHit[]> {
  let needle = q.toLowerCase().split(/\s+/).filter(Boolean).join(' ')
  if (!needle) return []
  needle = ALIASES[needle] ?? needle
  const all = await load(url)

  // Group by name so the result list shows variety: GNIS has 200-plus Bald
  // Mountains, and ranking purely by score buries everything else.
  const groups = new Map<string, [number, number, number, Row][]>()
  for (const r of all) {
    const s = score(needle, r.lower)
    if (s < 0) continue
    const key = r.lower
    const entry: [number, number, number, Row] = [s, CLASS_RANK[r.cls] ?? 99, r.name.length, r]
    const g = groups.get(key)
    if (g) g.push(entry)
    else groups.set(key, [entry])
  }

  const cmp = (a: [number, number, number, Row], b: [number, number, number, Row]): number =>
    a[0] - b[0] || a[1] - b[1] || a[2] - b[2]
  const ordered = [...groups.values()]
  for (const g of ordered) g.sort(cmp)
  ordered.sort((a, b) => cmp(a[0], b[0]))

  // Round-robin: every distinct name contributes its best hit before any name
  // contributes a second.
  const out: GazetteerHit[] = []
  const seen = new Set<string>()
  const deepest = ordered.reduce((m, g) => Math.max(m, g.length), 0)
  for (let depth = 0; depth < deepest; depth++) {
    for (const g of ordered) {
      if (depth >= g.length) continue
      const r = g[depth][3]
      const key = `${r.lower}|${r.lat.toFixed(2)}|${r.lon.toFixed(2)}`
      if (seen.has(key)) continue
      seen.add(key)
      const region = [r.county, r.state].filter(Boolean).join(', ')
      out.push({
        name: region ? `${r.name}, ${region}` : r.name,
        lat: r.lat,
        lon: r.lon,
        type: r.cls.toLowerCase(),
      })
      if (out.length >= limit) return out
    }
  }
  return out
}
